python/pyspark/ml/common.py

```diff
@@ -15,11 +15,15 @@
 # limitations under the License.
 #
 
+from typing import Any, Callable
+from pyspark.ml._typing import C, JavaObjectOrPickleDump
```

Comment:
This import should happen in TYPE_CHECKING block

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pyspark.ml._typing import C, JavaObjectOrPickleDump
```

Consequently, `JavaObjectOrPickleDump` and `C` have to be quoted when used (`"JavaObjectOrPickleDump"`).

Reply:
Ah OK. The `TYPE_CHECKING` block addresses the namespace pollution issue, I suppose.

But can you elaborate on why the type needs to be quoted? I understand that's for when the type is not known at that point in time (e.g. a self-referential type), but that isn't the case here.

Reply:
In general, whatever goes into `TYPE_CHECKING` is not imported during normal execution. So as is, all the names we use would be undefined when these scripts are imported.

[PEP 563](https://www.python.org/dev/peps/pep-0563/) introduces a concept of postponed evaluation, but were not ready to go there yet (for starters, we still didn't formally drop 3.6 support ‒ I am working on cleaning the code, then we have some components that might require code changes).